File: src/go/stack/isValidSerialization.go
package stack

func isValidSerialization(preorder string) bool {
	need := 1
	for idx := 0; idx < len(preorder); idx++ {
		if need == 0 {
			return false
		}
		if string(preorder[idx]) == "," {
			idx++
		} else if string(preorder[idx]) == "#" {
			need--
			idx++
		} else {
			for idx < len(preorder) && string(preorder[idx]) != "," {
				idx++
			}
			need += 1
		}
	}
	return need == 0
}

File: feishu/IsValidSerialization.java
import java.util.Stack;

/**
 * @author Zexho
 * @date 2022/4/23 10:42 PM
 */
public class IsValidSerialization {
    public boolean isValidSerialization(String preorder) {
        Stack<String> stack = new Stack<>();
        String[] strArr = preorder.split(",");
        int idx = 0;
        while(idx < strArr.length){
            String ch = strArr[idx];
            if(ch.equals("#")){
                while(!stack.isEmpty() && stack.peek().equals("#")){
                    stack.pop();
                    if(stack.isEmpty()){
                        return false;
                    }
                    stack.pop();
                }
            }
            stack.push(ch);
            idx += 1;
        }
        return stack.size() == 1 && stack.peek().equals("#");
    }
}
